File: scripts/core/systems/power.lua
local power = {}
local settings = require("scripts.game1.settings")
local draw = require("scripts.core.utils.draw")
local localization = require("scripts.core.utils.localization")

local PowerSystem = {}
PowerSystem.__index = PowerSystem

function PowerSystem.new(eng)
    local self = setmetatable({}, PowerSystem)
    self.eng = eng
    self.power = settings.MAX_POWER
    self.usage_level = 1
    self.is_dead = false
    self.dead_timer = 0.0
    self.freddy_music_playing = false
    return self
end

function PowerSystem:update(dt, door_usage, camera_open, vent_usage)
    vent_usage = vent_usage or 0
    if self.is_dead then
        self.dead_timer = self.dead_timer + dt
        return
    end

    self.usage_level = math.min(5, 1 + door_usage + (camera_open and 1 or 0) + vent_usage)
    local drain_mul = ({ 1.0, 1.55, 2.35, 3.4, 4.8 })[self.usage_level] or 1.0
    self.power = self.power - (settings.BASE_POWER_DRAIN * drain_mul) * dt
    if self.power <= 0 then
        self.power = 0
        self.is_dead = true
        self.dead_timer = 0.0
    end
end

function PowerSystem:draw(eng)
    local x, y = 24, settings.SCREEN_HEIGHT - 78
    local power_pct = math.max(0, math.floor(self.power))
    local color = (self.power > 20) and settings.POWER_COLOR or settings.POWER_LOW

    local text = string.format(localization.get_text("power_left"), power_pct)
    draw.text(eng, text, x, y, 18, settings.WHITE, 255, false)

    local uy = y + 28
    draw.text(eng, localization.get_text("usage"), x, uy, 16, settings.HUD_COLOR, 255, false)
    for i = 0, 4 do
        local ux = x + 78 + i * 16
        local c = (i < self.usage_level) and settings.MONITOR_GREEN or { 40, 40, 45 }
        eng:draw_rect(ux, uy + 4, 10, 16, c[1], c[2], c[3], 255)
        eng:draw_rect(ux, uy + 4, 10, 16, 75, 75, 80, 255, false)
    end
end

power.PowerSystem = PowerSystem
return power

File: src/systems/Power.cpp
#include "systems/Power.hpp"
#include "core/DrawUtils.hpp"
#include "core/Localization.hpp"
#include "game1/GameSettings.hpp"
#include <algorithm>
#include <cstdio>

namespace fnwf {

PowerSystem::PowerSystem() = default;

auto PowerSystem::update(double dt, int door_usage, bool camera_open) -> void {
    if (is_dead) {
        dead_timer += dt;
        return;
    }
    usage_level = std::min(5, 1 + door_usage + (camera_open ? 1 : 0));
    static const double drain_muls[] = {1.0, 1.55, 2.35, 3.4, 4.8};
    double mul = drain_muls[std::max(0, std::min(4, usage_level - 1))];
    power -= GameSettings::BASE_POWER_DRAIN * mul * dt;
    if (power <= 0.0) {
        power = 0.0;
        is_dead = true;
        dead_timer = 0.0;
    }
}

auto PowerSystem::draw(Engine& eng) -> void {
    auto x = 24;
    auto y = GameSettings::SCREEN_HEIGHT - 78;
    int pct = std::max(0, static_cast<int>(power));

    char buf[128];
    std::snprintf(buf, sizeof(buf), Localization::get_text("power_left").c_str(), pct);
    DrawUtils::text(eng, buf, x, y, 18, 255, 255, 255);

    auto uy = y + 28;
    DrawUtils::text(eng, Localization::get_text("usage"), x, uy, 16, 200, 200, 200);
    for (int i = 0; i < 5; ++i) {
        auto ux = x + 78 + i * 16;
        auto c = (i < usage_level) ? GameSettings::MONITOR_GREEN : Color{40, 40, 45};
        eng.draw_rect(ux, uy + 4, 10, 16, c.r, c.g, c.b);
        eng.draw_rect(ux, uy + 4, 10, 16, 75, 75, 80, 255, false);
    }
}

}  // namespace fnwf
